File: data-structures/Java/src/test/java/org/redquark/dsa/datastructures/lists/LinkedListTest.java
package org.redquark.dsa.datastructures.lists;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class LinkedListTest {

    private final LinkedList linkedList = new LinkedList();

    @BeforeEach
    void setUp() {
        linkedList.head = new Node(1);
        linkedList.head.next = new Node(2);
        linkedList.head.next.next = new Node(3);
        linkedList.head.next.next.next = new Node(4);
        linkedList.head.next.next.next.next = new Node(5);
    }

    @Test
    void testTraversal() {
        String expected = "1   2   3   4   5   ";
        assertEquals(expected, linkedList.traversal());
    }

    @Test
    void testGet() {
        assertEquals(3, linkedList.get(2));
        assertEquals(1, linkedList.get(0));
        assertEquals(-1, linkedList.get(9));
    }

    @Test
    void testInsertAtHead() {
        linkedList.insertAtHead(6);
        assertEquals(6, linkedList.get(0));
        linkedList.insertAtHead(7);
        assertEquals(7, linkedList.get(0));
    }

    @Test
    void testInsertAtTail() {
        linkedList.insertAtTail(6);
        assertEquals(6, linkedList.get(5));
        linkedList.insertAtTail(7);
        assertEquals(7, linkedList.get(6));
    }

    @Test
    void testInsertAtIndex() {
        linkedList.insertAtIndex(6, 1);
        assertEquals(2, linkedList.get(2));
        assertEquals(5, linkedList.get(5));
        linkedList.insertAtIndex(7, 4);
        assertEquals(4, linkedList.get(5));
    }

    @Test
    void testRemove() {
        linkedList.remove(linkedList.head);
        assertEquals(2, linkedList.get(0));
        linkedList.remove(linkedList.head.next.next);
        assertEquals(5, linkedList.get(2));
    }
}

File: data-structures/Python/lists/linked_list_test.py
import unittest
from lists import linked_list
from lists.linked_list import Node


class LinkedListTest(unittest.TestCase):

    @classmethod
    def setUp(cls):
        cls.head = Node(1)
        cls.head.next = Node(2)
        cls.head.next.next = Node(3)
        cls.head.next.next.next = Node(4)
        cls.head.next.next.next.next = Node(5)

    def test_traversal(self):
        expected = '1   2   3   4   5   '
        actual = linked_list.traversal(self)
        self.assertEqual(expected, actual)

    def test_get(self):
        self.assertEqual(2, linked_list.get(self, 1))
        self.assertEqual(5, linked_list.get(self, 4))
        self.assertEqual(-1, linked_list.get(self, 10))

    def test_insert_at_head(self):
        linked_list.insert_at_head(self, 6)
        self.assertEqual(6, linked_list.get(self, 0))
        self.assertEqual(1, linked_list.get(self, 1))
        linked_list.insert_at_head(self, 7)
        self.assertEqual(7, linked_list.get(self, 0))

    def test_insert_at_tail(self):
        linked_list.insert_at_tail(self, 6)
        self.assertEqual(6, linked_list.get(self, 5))
        linked_list.insert_at_tail(self, 7)
        self.assertEqual(7, linked_list.get(self, 6))

    def test_insert_at_position(self):
        linked_list.insert_at_index(self, 6, 2)
        self.assertEqual(3, linked_list.get(self, 3))
        linked_list.insert_at_index(self, 7, 1)
        self.assertEqual(6, linked_list.get(self, 3))

    def test_remove(self):
        linked_list.remove(self, self.head.next.next)
        self.assertEqual(4, linked_list.get(self, 2))
        linked_list.remove(self, self.head)
        self.assertEqual(2, linked_list.get(self, 0))


if __name__ == '__main__':
    unittest.main()
